Add index function and concat (hstack, vstack), linspace, arange and meshgrid

wolfv opened the pull request:
This is my approach to implement some classic numpy things.

Am happy about feedback. Do you think this is useful / a good approach?

JohanMabille commented:
We had a look early this morning with @SylvainCorlay, we really like your work. The code review will take some time, however the general approach is really good and my remarks should merely be about C++ semantic and style.

SylvainCorlay commented:
The windows build error is not legitimate. You can disregard it.

wolfv commented:
Hi both, 

thanks! :) You can consider this an initial draft and I'll spend some time today to fix some of the style issues I am sure you'd mention and add some documentation.

JohanMabille commented on include/xtensor/xbuilder.hpp:
```diff
@@ -69,6 +70,167 @@ namespace xt
+        template <class T, class U, class v_t = typename T::value_type>
+        struct concat_impl {
+            const T& m_A;
+            const U& m_B;
```
`m_A` and `m_B` should be renamed into `m_a` and `m_b`.

JohanMabille commented on include/xtensor/xbuilder.hpp:
```diff
@@ -69,6 +70,167 @@ namespace xt
+            }
+
+            value_type operator[](const xindex& idx) const {
+                return m_V[{idx[m_axis]}];
```
I don't understand why you use curly braces here. `return m_v(idx[m_axis]);` should be ok.

JohanMabille commented on include/xtensor/xbuilder.hpp:
```diff
@@ -69,6 +70,167 @@ namespace xt
+            const U& m_B;
+            using size_type = std::size_t;
+            using value_type = v_t;
+            std::size_t m_axis;
```
`using` definitions should appear before any member. If you define `size_type`, it should be used as the type of `m_axis`.

JohanMabille commented on include/xtensor/xbuilder.hpp:
```diff
@@ -69,6 +70,167 @@ namespace xt
+    }
+
+    template <class T>
+    inline auto arange(const T stop) noexcept
```
No need of const qualifier for arguments passed by value.

JohanMabille commented on include/xtensor/xscalar.hpp:
```diff
@@ -218,6 +219,12 @@ namespace xt
     }
 
     template <class T>
+    inline auto xscalar<T>::operator[](const xindex& idx) const noexcept -> const_reference
```
You can comment or remove the parameter name to avoid the annoying warning "parameter not used"

JohanMabille commented on test/test_xmath.cpp:
```diff
@@ -19,6 +19,12 @@ namespace xt
+        shape_type shape = {3, 2};
+        xarray<double> a(shape, 4.5);
+        EXPECT_EQ((pow(a, 2.f) + pow(a, 2.f))(0, 0), std::pow(a(0, 0), 2.f) + std::pow(a(0, 0), 2.f));
+    }
```
I don't understand why you add this test since the `pow` function is already tested after.

wolfv replied:
I added this test because I observed a problem with the missing `operator[](xindex)` on xscalars. But actually, this test didn't trigger the bug, so this is kind of an artifact.

JohanMabille commented:
I think you could add some unit tests since your patch is quite big. `xindex_function` could be tested in a similar way of `xfunction`, and then each function building on top of it can be tested very shortly.

wolfv commented on include/xtensor/xbuilder.hpp:
```diff
@@ -69,6 +70,167 @@ namespace xt
+            template <class... Args>
+            value_type operator()(Args... args) const {
+                std::vector<size_type> args_arr({static_cast<size_type>(args)...});
+                return m_V(args_arr[m_axis]);
```
Do you think the template version has many benefits over:

```
size_type args_arr[sizeof...(Args)] = {static_cast<size_type>(args)...};
return m_source(args_arr[m_axis]);
```

JohanMabille replied:
Actually no, since apply will allocate an array on the stack (if I remember well); your proposal is better because much simpler :)

SylvainCorlay commented:
One thing that would greatly improve performance is to make the `shape_type` of `xindex_function`  a template parameter to benefit from the faster std::array when possible.

The initializer from an `std::initializer_list` should result in an std::array rather than an `std::vector`

SylvainCorlay commented on include/xtensor/xbuilder.hpp:
```diff
@@ -69,6 +70,167 @@ namespace xt
+    }
+
+    template <class T, class U>
+    inline auto meshgrid(const T& X, const U& Y) noexcept
```
`numpy.meshgrid` takes an arbitrary number of arguments.

wolfv commented:
I fixed some things, but especially hstack and vstack proved to be a bit more difficult as numpy is actually extending shapes to make them fit (ie. np.vstack([1,2,3], [[1,2,3],[4,2,1]]) works. But for concatenate dimensions have to match up. 

I have also added quite a lot of tests.

Have to study now, but maybe after christmas we can finish this up :)

SylvainCorlay commented:
Regarding the meshgrid function, numpy's meshgrid takes any number of arrays.

There are two modes, `ij` and `xy`

```python
import numpy as np

x1, x2, x3, x4 = [1], [10, 20], [100, 200, 300], [1000, 2000, 3000, 4000]

ij = np.meshgrid(x1, x2, x3, x4, indexing='ij')
xy = np.meshgrid(x1, x2, x3, x4, indexing='xy')

print 'ij:', [m.shape for m in ij]
print 'xy:', [m.shape for m in xy]
```
prints

```python
ij: [(1, 2, 3, 4), (1, 2, 3, 4), (1, 2, 3, 4), (1, 2, 3, 4)]
xy: [(2, 1, 3, 4), (2, 1, 3, 4), (2, 1, 3, 4), (2, 1, 3, 4)]
```

meaning that 'xy' indexing simply **reverses the two first dimensions** of the result of 'ij' indexing.

`xy` is the default behavior of numpy, but I think that we should to `ij` by default in xtensor.

The reason for the nonsensical `xy` behavior is that it emulates Matlab.

It suggest we only do 'ij' by default, and have another mode that completely reverses the dimensions, and therefore would coincide with numpy in the case of two arguments.

Let's also think of what is the expected behavior with zero or one argument to meshgrid.

SylvainCorlay commented:
@wolfv would you consider splitting this into a first PR with only the `xindex_function` expression, `linspace`  and `arange` and another PR with the rest?

Besides, do you have other ideas for the class name than `xindex_function`? I feel that the concept is the right one, but we need a better name for it...

SylvainCorlay commented:
By the way we have made a 0.3.0 release. This would be really nice to make a 0.4 with this!

wolfv commented:
Cool! I'll try to polish it as much as i can over the next days then. 

Maybe something along the lines of `xaccess_function` or `xaccessor_function` makes more sense? Not sure if that naming is any better, though.

Btw. I also ported the random-intializer functions from the other PR to utilize `xindex_function` and do a lazy evaluation.

So far, there has not been a review of `xindex_function.hpp`, maybe one of you wants to do this?

SylvainCorlay commented:
I also thought of `xgenerator`, since it is also used by random generators.

wolfv commented:
great, makes total sense!

SylvainCorlay commented:
@wolv, do you want to close this and have a new PR for `concat / vstack / hstack / meshgrid`?

wolfv commented:
yes, that's what should happen. 
I'm a bit time-constrained now. 

Maybe I can get it done ~end of Februrary.
